src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,408 @@
+            EvictionMethod = EvictionMethod.Blocking
+        }
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{ToString()}")>]
+type CachedEntity<'Key, 'Value> =
+    val mutable Key: 'Key
+    val mutable Value: 'Value
+    val mutable AccessCount: int64
+    val mutable Node: LinkedListNode<CachedEntity<'Key, 'Value>>
+
+    new(key, value) =
+        {
+            Key = key
+            Value = value
+            AccessCount = 0L
+            Node = Unchecked.defaultof<_>
+        }
+
+    member this.WithNode() =
+        this.Node <- LinkedListNode(this)
+        this
+
+    member this.ReUse(key, value) =
+        this.Key <- key
+        this.Value <- value
+        this.AccessCount <- 0L
+        this
+
+    override this.ToString() = $"{this.Key}"
+
+type EntityPool<'Key, 'Value>(maximumCapacity, overCapacity: Event<_>) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
+    let mutable created = 0
+
+    member _.Acquire(key, value) =
+        match pool.TryTake() with
+        | true, entity -> entity.ReUse(key, value)
+        | _ ->
+            if Interlocked.Increment &created > maximumCapacity then
+                overCapacity.Trigger()
+
+            CachedEntity(key, value).WithNode()
+
+    member _.Reclaim(entity: CachedEntity<'Key, 'Value>) =
+        if pool.Count < maximumCapacity then
+            pool.Add(entity)
+
+type IEvictionQueue<'Key, 'Value> =
+    abstract member Add: CachedEntity<'Key, 'Value> * CachingStrategy -> unit
+    abstract member Update: CachedEntity<'Key, 'Value> -> unit
+    abstract member GetKeysToEvict: int -> 'Key[]
+    abstract member Remove: CachedEntity<'Key, 'Value> -> unit
+
+[<Sealed; NoComparison; NoEquality>]
+type EvictionQueue<'Key, 'Value>(strategy: CachingStrategy) =
+
+    let list = LinkedList<CachedEntity<'Key, 'Value>>()
+
+    interface IEvictionQueue<'Key, 'Value> with
+
+        member _.Add(entity: CachedEntity<'Key, 'Value>, strategy) =
+            lock list
+            <| fun () ->
+                if isNull entity.Node.List then
+                    match strategy with
+                    | CachingStrategy.LRU -> list.AddLast(entity.Node)
+                    | CachingStrategy.LFU -> list.AddLast(entity.Node)
+        // list.AddFirst(entity.Node)
+
+        member _.Update(entity: CachedEntity<'Key, 'Value>) =
+            lock list
+            <| fun () ->
+                Interlocked.Increment(&entity.AccessCount) |> ignore
+
+                let node = entity.Node
+
+                // Sync between store and the eviction queue is not atomic. It might be already evicted or not yet added.
+                if node.List = list then
+
+                    match strategy with
+                    | CachingStrategy.LRU ->
+                        // Just move this node to the end of the list.
+                        list.Remove(node)
+                        list.AddLast(node)
+                    | CachingStrategy.LFU ->
```

Comment:
If we are not going to use (therefore, not test/dogfood) `LFU` and only use either `no eviction` for OneOff compilation and `LRU` for IDE => WE can just live without `LFU` alltogether and get rid of it as an option and code supporting it.

(unless you are yet to evaluate which is better, in that case - I am sorry for bothering :) )

Reply:
Yes! It's LRU in the IDE, given the constant turnover while editing,